vault-plugins/flant_iam/model/group.go

```diff
@@ -255,3 +282,134 @@ func (r *GroupRepository) Sync(objID string, data []byte) error {
+}
+
+func (s subjectInTenantGroupIndexer) FromObject(raw interface{}) (bool, [][]byte, error) {
+	usersLabel := "Users"
+	serviceAccountsLabel := "ServiceAccounts"
+	groupsLabel := "Groups"
+	validSubjectFieldNames := map[string]struct{}{usersLabel: {}, serviceAccountsLabel: {}, groupsLabel: {}}
+	if _, valid := validSubjectFieldNames[s.subjectFieldName]; !valid {
+		return false, nil, fmt.Errorf("invalid subject_field_name: %s", s.subjectFieldName)
+	}
+	group, ok := raw.(*Group)
+	if !ok {
+		return false, nil, fmt.Errorf("format error: need Role type, actual passed %#v", raw)
+	}
+	result := [][]byte{}
+	tenantUUID := group.TenantUUID
+	switch s.subjectFieldName {
+	case usersLabel:
+		for i := range group.Users {
+			result = append(result, []byte(tenantUUID+group.Users[i]+"\x00"))
+		}
+	case serviceAccountsLabel:
+		for i := range group.ServiceAccounts {
+			result = append(result, []byte(tenantUUID+group.ServiceAccounts[i]+"\x00"))
+		}
+	case groupsLabel:
+		for i := range group.Groups {
+			result = append(result, []byte(tenantUUID+group.Groups[i]+"\x00"))
+		}
+	}
+	if len(result) == 0 {
+		return false, nil, nil
+	}
+	return true, result, nil
+}
+
+func (r *GroupRepository) findDirectParentGroupsByUserUUID(tenantUUID TenantUUID, userUUID UserUUID) (map[GroupUUID]struct{}, error) {
+	iter, err := r.db.Get(GroupType, UserInTenantGroupIndex, tenantUUID, userUUID)
+	if err != nil {
+		return nil, err
+	}
+
+	return extractGroupUUIDs(iter), nil
+}
+
+func (r *GroupRepository) findDirectParentGroupsByServiceAccountUUID(tenantUUID TenantUUID, serviceAccountUUID ServiceAccountUUID) (map[GroupUUID]struct{}, error) {
+	iter, err := r.db.Get(GroupType, ServiceAccountInTenantGroupIndex, tenantUUID, serviceAccountUUID)
+	if err != nil {
+		return nil, err
+	}
+	return extractGroupUUIDs(iter), nil
+}
+
+func (r *GroupRepository) findDirectParentGroupsByGroupUUID(tenantUUID TenantUUID, groupUUID GroupUUID) (map[GroupUUID]struct{}, error) {
+	iter, err := r.db.Get(GroupType, GroupInTenantGroupIndex, tenantUUID, groupUUID)
+	if err != nil {
+		return nil, err
+	}
+	return extractGroupUUIDs(iter), nil
+}
+
+// returns map with found parent uuids and originally passed uuids
+func (r *GroupRepository) findAllParentGroupsForGroupUUIDs(tenantUUID TenantUUID, groupUUIDs map[GroupUUID]struct{}) (map[GroupUUID]struct{}, error) {
+	resultGroupsSet := groupUUIDs
+	currentGroupsSet := groupUUIDs
+	for len(currentGroupsSet) != 0 {
+		nextSet := map[RoleName]struct{}{}
+		for currentGroupUUID := range currentGroupsSet {
+			candidates, err := r.findDirectParentGroupsByGroupUUID(tenantUUID, currentGroupUUID)
+			if err != nil {
+				return nil, err
+			}
+			for candidate := range candidates {
+				if _, found := resultGroupsSet[candidate]; !found {
+					resultGroupsSet[candidate] = struct{}{}
+					nextSet[candidate] = struct{}{}
+				}
+			}
+		}
+		currentGroupsSet = nextSet
+	}
+	return resultGroupsSet, nil
+}
```

Comment:
I see we use string sets a lot. I suggest we add more convenient wrapper for it to reduce struct{}{} boilerplate. Here is an example:

```

// Set is a set of strings. Not concurrency-safe.
type set map[string]struct{}

func (s set) Size() int {
	return len(s)
}

func (s set) Add(el string) {
	s[el] = struct{}{}
}

func newSet(elems ...string) set {
	set := set{}
	for _, el := range elems {
		set.Add(el)
	}
	return set
}

func (s set) Has(el string) bool {
	_, has := s[el]
	return has 
}
```

Reply:
We will lost  GroupUUID in signatures in this case...